### PR TITLE
New Assign Gear

### PR DESCRIPTION
**Pull request by PabstMirror**
- Should hopefully fix that no-uniform bug
 - Move more code out of missions

Instead of adding weapons and items to a unit, we build up a unitLoadout array from the configs, cache it and then apply it all at once with `setUnitLoadout`. Once cached it's about 7x faster compared to the old system.

Seems to be working.  I have a mission with this code, which should be a good test at full player counts.  Vehicle assign is unchanged from bwmf version.

### CODE REVIEW

File: addons/assignGear/functions/fnc_assignGearMan.sqf [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        TRACE_1("compiling new",_loadoutKey);
+        _loadoutArray = [_path] call FUNC(getLoadoutFromConfig);
+        GVAR(loadoutCache) setVariable [_loadoutKey, _loadoutArray];
+    };
+
+    _unit setUnitLoadout _loadoutArray;
+
+    if (isText (_path >> "init")) then {
+        TRACE_1("calling init code",getText (_path >> "init"));
+        _unit call compile ("this = _this;"+ getText (_path >> "init"));
```
Comment (AACO) [inline addons/assignGear/functions/fnc_assignGearMan.sqf]: Just curious, are we planning on including init code here, or is it just a bonus feature?

File: addons/assignGear/functions/fnc_getLoadoutFromConfig.sqf [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+    _vestArray = [_vest, []];
+    _containersArray set [1, [_vestSpace, _vestArray]];
+};
+if (!(_configBackpack isEqualTo [])) then {
+    private _backpack = selectRandom _configBackpack;
+    private _backpackSpace = getNumber (configFile >> "CfgVehicles" >> _backpack >> "maximumLoad");
+    TRACE_2("",_backpack,_backpackSpace);
+    _backpackArray = [_backpack, []];
+    _containersArray set [2, [_backpackSpace, _backpackArray]];
+};
```
Comment (AACO) [inline addons/assignGear/functions/fnc_getLoadoutFromConfig.sqf]: Should we do checks above to ensure the entered class names are valid?

File: addons/assignGear/functions/fnc_assignGearMan.sqf (hunk 1 above)
Comment (PabstMirror) [inline addons/assignGear/functions/fnc_assignGearMan.sqf]: Just a bonus, I think waffle added this

**Comment by AACO** [comment]
Are we removing the gear logger? I feel like that's a super helpful tool to mission makers/people managing loadouts.

**Comment by AACO** [comment]
Maybe use some defines in getLoadoutFromConfig to help give meaning to some of the magic numbers?

File: addons/assignGear/functions/fnc_assignGearMan.sqf [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    private _loadoutArray = GVAR(loadoutCache) getVariable _loadoutKey;
+
+    if (isNil "_loadoutArray") then {
+        TRACE_1("compiling new",_loadoutKey);
+        _loadoutArray = [_path] call FUNC(getLoadoutFromConfig);
+        GVAR(loadoutCache) setVariable [_loadoutKey, _loadoutArray];
+    };
+
+    _unit setUnitLoadout _loadoutArray;
```
Comment (AACO) [inline addons/assignGear/functions/fnc_assignGearMan.sqf]: Do we know if this is a global command or not? The wiki is less than helpful.

Comment (PabstMirror) [inline addons/assignGear/functions/fnc_assignGearMan.sqf]: It works on remote units; the wiki says
>	This command is not final and might be changed in a near future.
But who knows what BIS will do

File: addons/assignGear/config.cpp [hunk 4]
```diff
@@ -0,0 +1,15 @@
+#include "script_component.hpp"
+
+class CfgPatches {
+    class ADDON {
+        units[] = {};
+        weapons[] = {};
+        requiredVersion = REQUIRED_VERSION;
```
Comment (AACO) [inline addons/assignGear/config.cpp]: Required version should be bumped to 1.58 in core

**Comment by PabstMirror** [comment]
1.60 should make this viable: on hold until then

**Comment by AACO** [comment]
Just curious, what is 1.60 adding that'll help us out here?

**Comment by PabstMirror** [comment]
>Tweaked: The setUnitLoadout script command is now supported in the multiplayer environment

https://forums.bistudio.com/topic/140837-development-branch-changelog/page-39

**Comment by PabstMirror** [comment]
Looks like the totally broke this command in 1.60 - https://feedback.bistudio.com/T116980

**Comment by AACO** [comment]
Fuuuuuuuccckkkkk

**Comment by PabstMirror** [comment]
Looks like we're back in business with 1.62

File: addons/assignGear/CfgEventHandlers.hpp [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    };
+};
+class Extended_PreInit_EventHandlers {
+    class ADDON {
+        init = QUOTE(call COMPILE_FILE(XEH_preInit));
+    };
+};
+
+
+class Extended_Init_EventHandlers {
```
Comment (AACO) [inline addons/assignGear/CfgEventHandlers.hpp]: Should we add these only if the system is enabled? (CBA lets you run it for all spawned units when adding a new XEH)

File: addons/assignGear/functions/fnc_getLoadoutFromConfig.sqf [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+private _configBackpack = getArray (_path >> "backpack");
+private _configBackpackItems = getArray (_path >> "backpackItems");
+private _configWeapons = getArray (_path >> "weapons");
+private _configLaunchers = getArray (_path >> "launchers");
+private _configHandguns = getArray (_path >> "handguns");
+private _configMagazines = getArray (_path >> "magazines");
+private _configItems = getArray (_path >> "items");
+private _configLinkedItems = getArray (_path >> "linkedItems");
+private _configAttachments = getArray (_path >> "attachments");
+private _configSecondaryAttachments = getArray (_path >> "secondaryAttachments");
```
Comment (AACO) [inline addons/assignGear/functions/fnc_getLoadoutFromConfig.sqf]: We should handle handgun attachments separately as well.

File: addons/assignGear/functions/fnc_getLoadoutFromConfig.sqf [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+    _containersArray set [2, [_backpackSpace, _backpackArray]];
+};
+
+TRACE_1("containers",_containersArray);
+
+private _headgear = if (_configHeadgear isEqualTo []) then {""} else {selectRandom _configHeadgear};
+private _binocularArray = [];
+private _assignedItems = ["", "", "", "", "", ""];
+
+private _fnc_addItemsToContainer = {
```
Comment (AACO) [inline addons/assignGear/functions/fnc_getLoadoutFromConfig.sqf]: Should these private funcs be externalized? This function will be called a lot, and we're wasting some IO re-deffing them each call

File: addons/assignGear/CfgEventHandlers.hpp [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    };
+    class Helicopter {
+        class ADDON { serverInit  = QUOTE([ARR_2((_this select 0), 'Helicopter')] call FUNC(assignGearVehicle);); };
+    };
+    class Plane {
+        class ADDON { serverInit  = QUOTE([ARR_2((_this select 0), 'Plane')] call FUNC(assignGearVehicle);); };
+    };
+    class Ship_F {
+        class ADDON { serverInit  = QUOTE([ARR_2((_this select 0), 'Ship_F')] call FUNC(assignGearVehicle);); };
+    };
```
Comment (AACO) [inline addons/assignGear/CfgEventHandlers.hpp]: Do we just want to use https://github.com/CBATeam/CBA_A3/blob/master/addons/xeh/fnc_addClassEventHandler.sqf and use the apply retroactively flag after we determine it's enabled or not?

**Pull request merged**